File: src/games/truth-or-dare/logic.ts
// src/games/truth-or-dare/logic.ts — PURE logic. No clock/RNG/IO; seeds via createInitialState / action payloads.
import type { ColorToken, GameConfig, GameStateBase } from '../../sdk/types';
import type { DeckState } from '../../engine/deck';
import type { ScoreState } from '../../engine/scoring';
import type { RevealGateState } from '../../engine/revealGate';
import * as deck from '../../engine/deck';
import * as scoring from '../../engine/scoring';
import * as revealGate from '../../engine/revealGate';
import { pick, deriveSeed } from '../../engine/rng';
import { asPlayerId } from '../../engine/ids';
import { PROMPT_BY_ID, getPool } from './content';
import type { Intensity, PromptItem, PromptKind } from './content';
import { readOptions } from './config';
import type { EndType, ToDOptions } from './config';

export type ToDPhase = 'idle' | 'choosing' | 'revealing' | 'resolving' | 'gameOver' | 'error';
export type Outcome = 'done' | 'skip';

export interface TurnRecord {
  turnIndex: number;
  playerId: string;
  kind: PromptKind;
  promptId: string;
  intensity: Intensity;
  outcome: Outcome;
  pointsDelta: number;
}

export interface ToDState extends GameStateBase {
  phase: ToDPhase;
  options: ToDOptions;
  playerIds: string[];
  playerNames: Record<string, string>;
  playerColors: Record<string, ColorToken | undefined>;
  activePlayerId: string | null;
  lastPlayerId: string | null;
  cursor: number; // sequential rotation pointer
  currentKind: PromptKind | null;
  currentPromptId: string | null;
  truthDeck: DeckState<string>;
  dareDeck: DeckState<string>;
  scores: ScoreState;
  reveal: RevealGateState | null;
  turnIndex: number;
  roundIndex: number;
  spinSerial: number;
  history: TurnRecord[];
  endReason: EndType | null;
  errorCode: 'EMPTY_POOL' | 'NOT_ENOUGH_PLAYERS' | null;
}

export type ToDAction =
  | { type: 'SPIN'; seed: number }
  | { type: 'NEXT_PLAYER' }
  | { type: 'CHOOSE'; kind: PromptKind; seed: number }
  | { type: 'REVEAL' }
  | { type: 'RESOLVE'; outcome: Outcome }
  | { type: 'REDRAW'; seed: number }
  | { type: 'END_GAME' };

export function pickSpinTarget(
  playerIds: string[],
  lastPlayerId: string | null,
  avoid: boolean,
  seed: number,
): string {
  let candidates = playerIds;
  if (avoid && playerIds.length > 1 && lastPlayerId) {
    const filtered = playerIds.filter((id) => id !== lastPlayerId);
    if (filtered.length > 0) candidates = filtered;
  }
  return pick(candidates, seed);
}

function shouldGate(options: ToDOptions, prompt: PromptItem): boolean {
  if (options.privateReveal === 'always') return true;
  if (options.privateReveal === 'spicyOnly') return prompt.intensity === 'spicy';
  return false;
}

function computeDelta(options: ToDOptions, kind: PromptKind, outcome: Outcome): number {
  if (options.scoringMode !== 'points') return 0;
  if (outcome === 'skip') return options.pointsForSkip;
  return kind === 'dare' ? options.pointsForDare : options.pointsForTruth;
}

function evalEnd(options: ToDOptions, scores: ScoreState, roundIndex: number): EndType | null {
  if (options.endType === 'rounds') return roundIndex >= options.endValue ? 'rounds' : null;
  if (options.endType === 'target') {
    if (options.scoringMode !== 'points') return null;
    const max = Math.max(0, ...Object.values(scores.totals));
    return max >= options.endValue ? 'target' : null;
  }
  return null;
}

function drawPrompt(
  d: DeckState<string>,
  seed: number,
): { promptId: string | null; deck: DeckState<string> } {
  const r = deck.draw(d, 1, seed);
  const promptId = r.drawn[0] ?? null;
  const next = promptId !== null ? deck.discard(r.deck, promptId) : r.deck;
  return { promptId, deck: next };
}

export function createInitialState(config: GameConfig, seed: number): ToDState {
  const options = readOptions(config);
  const playerIds = config.players.map((p) => p.id as string);
  const playerNames: Record<string, string> = {};
  const playerColors: Record<string, ColorToken | undefined> = {};
  config.players.forEach((p) => {
    playerNames[p.id] = p.name;
    playerColors[p.id] = p.color;
  });

  const pc = Math.max(2, playerIds.length);
  const truthPool = getPool({ kind: 'truth', intensities: options.intensities, playerCount: pc });
  const darePool = getPool({ kind: 'dare', intensities: options.intensities, playerCount: pc });

  const errorCode: ToDState['errorCode'] =
    playerIds.length < 2
      ? 'NOT_ENOUGH_PLAYERS'
      : truthPool.length === 0 || darePool.length === 0
        ? 'EMPTY_POOL'
        : null;

  return {
    v: 1,
    phase: errorCode ? 'error' : 'idle',
    finished: false,
    options,
    playerIds,
    playerNames,
    playerColors,
    activePlayerId: null,
    lastPlayerId: null,
    cursor: -1,
    currentKind: null,
    currentPromptId: null,
    truthDeck: deck.create(truthPool.map((p) => p.id), seed),
    dareDeck: deck.create(darePool.map((p) => p.id), deriveSeed(seed, 1)),
    scores: scoring.create(playerIds),
    reveal: null,
    turnIndex: 0,
    roundIndex: 0,
    spinSerial: 0,
    history: [],
    endReason: null,
    errorCode,
  };
}

function startTurn(s: ToDState, target: string): ToDState {
  return {
    ...s,
    phase: 'choosing',
    lastPlayerId: s.activePlayerId,
    activePlayerId: target,
    currentKind: null,
    currentPromptId: null,
    reveal: null,
    spinSerial: s.spinSerial + 1,
  };
}

function drawInto(s: ToDState, kind: PromptKind, seed: number): ToDState {
  const deckKey = kind === 'truth' ? 'truthDeck' : 'dareDeck';
  const { promptId, deck: nextDeck } = drawPrompt(s[deckKey], seed);
  if (promptId === null) return s;
  const prompt = PROMPT_BY_ID[promptId];
  const gate = shouldGate(s.options, prompt);
  return {
    ...s,
    currentKind: kind,
    currentPromptId: promptId,
    [deckKey]: nextDeck,
    reveal: gate ? revealGate.init([asPlayerId(s.activePlayerId ?? '')]) : null,
    phase: gate ? 'revealing' : 'resolving',
  };
}

export function reducer(state: ToDState, action: ToDAction): ToDState {
  const s = state;
  switch (action.type) {
    case 'SPIN': {
      if (s.phase !== 'idle' || s.options.selectionMode !== 'spinner') return s;
      const target = pickSpinTarget(
        s.playerIds,
        s.lastPlayerId,
        s.options.avoidImmediateRepeat,
        action.seed,
      );
      return startTurn(s, target);
    }
    case 'NEXT_PLAYER': {
      if (s.phase !== 'idle' || s.options.selectionMode !== 'sequential') return s;
      const cursor = (s.cursor + 1) % s.playerIds.length;
      return { ...startTurn(s, s.playerIds[cursor]), cursor };
    }
    case 'CHOOSE': {
      if (s.phase !== 'choosing' || !s.activePlayerId) return s;
      return drawInto(s, action.kind, action.seed);
    }
    case 'REVEAL': {
      if (s.phase !== 'revealing' || !s.reveal) return s;
      return { ...s, reveal: revealGate.reveal(s.reveal), phase: 'resolving' };
    }
    case 'REDRAW': {
      if ((s.phase !== 'revealing' && s.phase !== 'resolving') || !s.currentKind) return s;
      return drawInto(s, s.currentKind, action.seed);
    }
    case 'RESOLVE': {
      const okPhase =
        s.phase === 'resolving' || (s.phase === 'revealing' && action.outcome === 'skip');
      if (!okPhase || !s.activePlayerId || !s.currentKind || !s.currentPromptId) return s;
      const prompt = PROMPT_BY_ID[s.currentPromptId];
      const delta = computeDelta(s.options, s.currentKind, action.outcome);
      const scores = scoring.add(s.scores, s.activePlayerId, delta, action.outcome, 0);
      const record: TurnRecord = {
        turnIndex: s.turnIndex,
        playerId: s.activePlayerId,
        kind: s.currentKind,
        promptId: s.currentPromptId,
        intensity: prompt.intensity,
        outcome: action.outcome,
        pointsDelta: delta,
      };
      const turnIndex = s.turnIndex + 1;
      const roundIndex =
        turnIndex % s.playerIds.length === 0 ? s.roundIndex + 1 : s.roundIndex;
      const endReason = evalEnd(s.options, scores, roundIndex);
      return {
        ...s,
        scores,
        history: [...s.history, record],
        turnIndex,
        roundIndex,
        currentKind: null,
        currentPromptId: null,
        reveal: null,
        phase: endReason ? 'gameOver' : 'idle',
        endReason,
        finished: !!endReason,
      };
    }
    case 'END_GAME': {
      if (s.phase === 'gameOver') return s;
      return { ...s, phase: 'gameOver', endReason: 'endless', finished: true };
    }
    default:
      return s;
  }
}

/* ─────────────────────────  Pure selectors  ───────────────────────── */

export interface ToDStanding {
  id: string;
  score: number;
  rank: number;
}

export function standings(s: ToDState): ToDStanding[] {
  const order = new Map(s.playerIds.map((id, i) => [id, i]));
  const sorted = [...s.playerIds].sort(
    (a, b) =>
      scoring.total(s.scores, b) - scoring.total(s.scores, a) ||
      (order.get(a)! - order.get(b)!),
  );
  let rank = 0;
  let prev = NaN;
  return sorted.map((id, i) => {
    const sc = scoring.total(s.scores, id);
    if (sc !== prev) {
      rank = i + 1;
      prev = sc;
    }
    return { id, score: sc, rank };
  });
}

export function computeWinners(s: ToDState): string[] {
  if (s.options.scoringMode !== 'points') return [];
  return standings(s)
    .filter((r) => r.rank === 1)
    .map((r) => r.id);
}

export const nextSequentialId = (s: ToDState): string =>
  s.playerIds[(s.cursor + 1) % s.playerIds.length] ?? '';
export const promptRevealed = (s: ToDState): boolean =>
  s.phase === 'resolving' || (!!s.reveal && revealGate.holder(s.reveal) === undefined);

File: android-native/app/src/main/java/com/gamenight/party/game/truth-or-dare/Logic.kt
package com.gamenight.party.game.truthordare

import com.gamenight.party.engine.DeckState
import com.gamenight.party.engine.RevealGateState
import com.gamenight.party.engine.Rng
import com.gamenight.party.engine.ScoreState
import com.gamenight.party.engine.add
import com.gamenight.party.engine.asPlayerId
import com.gamenight.party.engine.create
import com.gamenight.party.engine.deriveSeed
import com.gamenight.party.engine.discard
import com.gamenight.party.engine.draw
import com.gamenight.party.engine.holder
import com.gamenight.party.engine.init
import com.gamenight.party.engine.reveal
import com.gamenight.party.engine.total
import com.gamenight.party.model.ColorToken

/**
 * PURE logic for "Truth or Dare" — a faithful Kotlin port of src/games/truth-or-dare/logic.ts.
 *
 * No clock / RNG / IO here: seeds arrive on the actions (or [createInitialState]); the Compose layer
 * owns the impure boundary (sound, haptics, seed generation). The web reducer reads a module-global
 * `PROMPT_BY_ID` for a prompt's intensity; to keep this reducer self-contained and pure, that lookup
 * is captured into [ToDState.intensityById] at init time instead.
 */

enum class ToDPhase { IDLE, CHOOSING, REVEALING, RESOLVING, GAME_OVER, ERROR }

enum class Outcome { DONE, SKIP }

enum class ToDErrorCode { EMPTY_POOL, NOT_ENOUGH_PLAYERS }

/** One completed turn, appended to [ToDState.history] (src `TurnRecord`). */
data class TurnRecord(
    val turnIndex: Int,
    val playerId: String,
    val kind: PromptKind,
    val promptId: String,
    val intensity: Intensity,
    val outcome: Outcome,
    val pointsDelta: Int,
)

/** The whole game state (src `ToDState`, which extends `GameStateBase` with `v` + `finished`). */
data class ToDState(
    val v: Int = 1,
    val finished: Boolean = false,
    val phase: ToDPhase,
    val options: ToDOptions,
    val playerIds: List<String>,
    val playerNames: Map<String, String>,
    val playerColors: Map<String, ColorToken?>,
    val activePlayerId: String? = null,
    val lastPlayerId: String? = null,
    /** sequential rotation pointer. */
    val cursor: Int = -1,
    val currentKind: PromptKind? = null,
    val currentPromptId: String? = null,
    val truthDeck: DeckState<String>,
    val dareDeck: DeckState<String>,
    val scores: ScoreState,
    val reveal: RevealGateState? = null,
    val turnIndex: Int = 0,
    val roundIndex: Int = 0,
    val spinSerial: Int = 0,
    val history: List<TurnRecord> = emptyList(),
    val endReason: EndType? = null,
    val errorCode: ToDErrorCode? = null,
    /** Localized mirror of the web's `PROMPT_BY_ID[id].intensity`, covering both active pools. */
    val intensityById: Map<String, Intensity> = emptyMap(),
)

/** src `ToDAction`. */
sealed interface ToDAction {
    data class Spin(val seed: Int) : ToDAction
    data object NextPlayer : ToDAction
    data class Choose(val kind: PromptKind, val seed: Int) : ToDAction
    data object Reveal : ToDAction
    data class Resolve(val outcome: Outcome) : ToDAction
    data class Redraw(val seed: Int) : ToDAction
    data object EndGame : ToDAction
}

/** src `pickSpinTarget`. Deterministic given the seed; avoids the last player when possible. */
fun pickSpinTarget(
    playerIds: List<String>,
    lastPlayerId: String?,
    avoid: Boolean,
    seed: Int,
): String {
    var candidates = playerIds
    if (avoid && playerIds.size > 1 && lastPlayerId != null) {
        val filtered = playerIds.filter { it != lastPlayerId }
        if (filtered.isNotEmpty()) candidates = filtered
    }
    return Rng(seed).pick(candidates)
}

private fun shouldGate(options: ToDOptions, intensity: Intensity): Boolean = when (options.privateReveal) {
    PrivateReveal.ALWAYS -> true
    PrivateReveal.SPICY_ONLY -> intensity == Intensity.SPICY
    PrivateReveal.NEVER -> false
}

private fun computeDelta(options: ToDOptions, kind: PromptKind, outcome: Outcome): Int {
    if (options.scoringMode != ScoringMode.POINTS) return 0
    if (outcome == Outcome.SKIP) return options.pointsForSkip
    return if (kind == PromptKind.DARE) options.pointsForDare else options.pointsForTruth
}

private fun evalEnd(options: ToDOptions, scores: ScoreState, roundIndex: Int): EndType? = when (options.endType) {
    EndType.ROUNDS -> if (roundIndex >= options.endValue) EndType.ROUNDS else null
    EndType.TARGET -> {
        if (options.scoringMode != ScoringMode.POINTS) {
            null
        } else {
            val max = (scores.totals.values + 0).maxOrNull() ?: 0
            if (max >= options.endValue) EndType.TARGET else null
        }
    }
    EndType.ENDLESS -> null
}

/** Draw one prompt id off [d], discarding it back so the deck recycles after exhaustion. */
private fun drawPrompt(d: DeckState<String>, seed: Int): Pair<String?, DeckState<String>> {
    val r = draw(d, 1, seed)
    val promptId = r.drawn.firstOrNull()
    val next = if (promptId != null) discard(r.deck, promptId) else r.deck
    return promptId to next
}

/** src `createInitialState`. Content is fed in (web reads it from a module global). */
fun createInitialState(content: ToDContent, config: ToDConfig, seed: Int): ToDState {
    val options = readOptions(config)
    val playerIds = config.players.map { it.id }
    val playerNames = config.players.associate { it.id to it.name }
    val playerColors = config.players.associate { it.id to it.color }

    val pc = maxOf(2, playerIds.size)
    val truthPool = content.getPool(PromptKind.TRUTH, options.intensities, pc)
    val darePool = content.getPool(PromptKind.DARE, options.intensities, pc)

    val errorCode = when {
        playerIds.size < 2 -> ToDErrorCode.NOT_ENOUGH_PLAYERS
        truthPool.isEmpty() || darePool.isEmpty() -> ToDErrorCode.EMPTY_POOL
        else -> null
    }

    val intensityById = (truthPool + darePool).associate { it.id to it.intensity }

    return ToDState(
        v = 1,
        phase = if (errorCode != null) ToDPhase.ERROR else ToDPhase.IDLE,
        finished = false,
        options = options,
        playerIds = playerIds,
        playerNames = playerNames,
        playerColors = playerColors,
        activePlayerId = null,
        lastPlayerId = null,
        cursor = -1,
        currentKind = null,
        currentPromptId = null,
        truthDeck = create(truthPool.map { it.id }, seed),
        dareDeck = create(darePool.map { it.id }, deriveSeed(seed, 1)),
        scores = create(playerIds),
        reveal = null,
        turnIndex = 0,
        roundIndex = 0,
        spinSerial = 0,
        history = emptyList(),
        endReason = null,
        errorCode = errorCode,
        intensityById = intensityById,
    )
}

private fun startTurn(s: ToDState, target: String): ToDState = s.copy(
    phase = ToDPhase.CHOOSING,
    lastPlayerId = s.activePlayerId,
    activePlayerId = target,
    currentKind = null,
    currentPromptId = null,
    reveal = null,
    spinSerial = s.spinSerial + 1,
)

private fun drawInto(s: ToDState, kind: PromptKind, seed: Int): ToDState {
    val isTruth = kind == PromptKind.TRUTH
    val sourceDeck = if (isTruth) s.truthDeck else s.dareDeck
    val (promptId, nextDeck) = drawPrompt(sourceDeck, seed)
    if (promptId == null) return s
    val intensity = s.intensityById[promptId] ?: Intensity.MILD
    val gate = shouldGate(s.options, intensity)
    return s.copy(
        currentKind = kind,
        currentPromptId = promptId,
        truthDeck = if (isTruth) nextDeck else s.truthDeck,
        dareDeck = if (isTruth) s.dareDeck else nextDeck,
        reveal = if (gate) init(listOf(asPlayerId(s.activePlayerId ?: ""))) else null,
        phase = if (gate) ToDPhase.REVEALING else ToDPhase.RESOLVING,
    )
}

/** src `reducer`. Pure: same input always yields the same output. */
fun reducer(state: ToDState, action: ToDAction): ToDState {
    val s = state
    return when (action) {
        is ToDAction.Spin -> {
            if (s.phase != ToDPhase.IDLE ||
                (s.options.selectionMode != SelectionMode.SPINNER && s.options.selectionMode != SelectionMode.BOTTLE)
            ) {
                s
            } else {
                val target = pickSpinTarget(s.playerIds, s.lastPlayerId, s.options.avoidImmediateRepeat, action.seed)
                startTurn(s, target)
            }
        }

        is ToDAction.NextPlayer -> {
            if (s.phase != ToDPhase.IDLE || s.options.selectionMode != SelectionMode.SEQUENTIAL || s.playerIds.isEmpty()) {
                s
            } else {
                val cursor = (s.cursor + 1) % s.playerIds.size
                startTurn(s, s.playerIds[cursor]).copy(cursor = cursor)
            }
        }

        is ToDAction.Choose -> {
            if (s.phase != ToDPhase.CHOOSING || s.activePlayerId == null) s
            else drawInto(s, action.kind, action.seed)
        }

        is ToDAction.Reveal -> {
            val rev = s.reveal
            if (s.phase != ToDPhase.REVEALING || rev == null) s
            else s.copy(reveal = reveal(rev), phase = ToDPhase.RESOLVING)
        }

        is ToDAction.Redraw -> {
            val kind = s.currentKind
            if ((s.phase != ToDPhase.REVEALING && s.phase != ToDPhase.RESOLVING) || kind == null) s
            else drawInto(s, kind, action.seed)
        }

        is ToDAction.Resolve -> {
            val active = s.activePlayerId
            val kind = s.currentKind
            val promptId = s.currentPromptId
            val okPhase =
                s.phase == ToDPhase.RESOLVING || (s.phase == ToDPhase.REVEALING && action.outcome == Outcome.SKIP)
            if (!okPhase || active == null || kind == null || promptId == null) {
                s
            } else {
                val intensity = s.intensityById[promptId] ?: Intensity.MILD
                val delta = computeDelta(s.options, kind, action.outcome)
                val reason = if (action.outcome == Outcome.SKIP) "skip" else "done"
                val scores = add(s.scores, active, delta, reason, 0L)
                val record = TurnRecord(
                    turnIndex = s.turnIndex,
                    playerId = active,
                    kind = kind,
                    promptId = promptId,
                    intensity = intensity,
                    outcome = action.outcome,
                    pointsDelta = delta,
                )
                val turnIndex = s.turnIndex + 1
                val roundIndex = if (turnIndex % s.playerIds.size == 0) s.roundIndex + 1 else s.roundIndex
                val endReason = evalEnd(s.options, scores, roundIndex)
                s.copy(
                    scores = scores,
                    history = s.history + record,
                    turnIndex = turnIndex,
                    roundIndex = roundIndex,
                    currentKind = null,
                    currentPromptId = null,
                    reveal = null,
                    phase = if (endReason != null) ToDPhase.GAME_OVER else ToDPhase.IDLE,
                    endReason = endReason,
                    finished = endReason != null,
                )
            }
        }

        is ToDAction.EndGame -> {
            if (s.phase == ToDPhase.GAME_OVER) s
            else s.copy(phase = ToDPhase.GAME_OVER, endReason = EndType.ENDLESS, finished = true)
        }
    }
}

/* ─────────────────────────  Pure selectors  ───────────────────────── */

/** One ranked row (src `ToDStanding`). */
data class ToDStanding(val id: String, val score: Int, val rank: Int)

/** src `standings` — total desc, original seat order as the tiebreak, with dense-ish ranking. */
fun standings(s: ToDState): List<ToDStanding> {
    val order = s.playerIds.withIndex().associate { (i, id) -> id to i }
    val sorted = s.playerIds.sortedWith(
        compareByDescending<String> { total(s.scores, it) }.thenBy { order[it] ?: 0 },
    )
    var rank = 0
    var prev: Int? = null
    return sorted.mapIndexed { i, id ->
        val sc = total(s.scores, id)
        if (sc != prev) {
            rank = i + 1
            prev = sc
        }
        ToDStanding(id, sc, rank)
    }
}

/** src `computeWinners` — only meaningful in points mode (everyone tied at rank 1). */
fun computeWinners(s: ToDState): List<String> {
    if (s.options.scoringMode != ScoringMode.POINTS) return emptyList()
    return standings(s).filter { it.rank == 1 }.map { it.id }
}

/** src `nextSequentialId`. */
fun nextSequentialId(s: ToDState): String =
    if (s.playerIds.isEmpty()) "" else s.playerIds[(s.cursor + 1) % s.playerIds.size]

/** src `promptRevealed`. */
fun promptRevealed(s: ToDState): Boolean {
    val rev = s.reveal
    return s.phase == ToDPhase.RESOLVING || (rev != null && holder(rev) == null)
}
